proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+
+        X' = X \ { qual ↦ orig | qual ↦ orig ∈ X
+                               , qual ↦ orig' ∈ X
+                               , orig /= orig' }
+
+      Now, ``X'`` is our final export QEnv.
+
+   #. **Examples**. For instance in
+
+      ::
+
+        module M (module qualified X (f)) where
+
+          module qualified X where
+            f = …
+            g = …
+
+      the module ``M`` exports ``{X.f ↦ (M, X.f)}``. And in
+
+      ::
+
+       module M (module qualified X) where
+
+         module qualified X where
+           f = …
+           g = …
+
+      the module ``M`` exports ``{X.f ↦ (M, X.f), X.g ↦ (M, X.g)}``. And in
+
+      ::
+
+         module X where
+           module Y where
+             f = …
+             g = …
+
+         module M (module X) where
+           import X
+
+      the export environment of ``M`` is ``{f ↦ (X, Y.f), g ↦ (x, Y.g)}``.
+
+#. **Local Modules**
+
+   A. Introduce a new declaration form
+      to declare a new module called a *local module*. Here is the BNF::
+
+        topdecl ::= ... | 'module' [ 'qualified' ] modids [ exports ] 'where' '{' topdecls '}'
+
+      This declaration form is allowed only with ``-XLocalModules``.
+
+      The ``module`` keyword in a local module declaration must not be the first
+      lexeme in a file. (This is to avoid ambiguity with top-level,
+      file-sized modules.)
+
+      Note that this new form is part of ``topdecl``, meaning that it cannot appear
+      in a ``let`` or ``where``. However, because a local module contains ``topdecl``\ s,
+      local modules may be arbitrarily nested.
+
+   #. There are now two kinds of modules: top-level modules and local
+      modules. Local modules lack import statements, but can define all
+      the range of entities definable in a top-level module.
+
+      The term *module* can refer to either a top-level or a local module.
+
+   #. The ambient QEnv within the local module extends the ambient QEnv
+      of the enclosing module with its local definitions. This means that
+      all qualified names in scope in the enclosing module are in scope in the
+      local module.
+
+   #. Definitions in a local module may be mutually recursive with definitions
+      in other local modules or outside of any local module. That is, local
+      modules influence scoping only, but not type-checking or dependency
+      (which remain constrained by compilation units, as they are today).
+
+   #. Let the ambient QEnv of the local module (call it ``M``) be ``E``.
+      The export QEnv
+      of ``M`` is the result of interpreting ``exports`` (if given)
+      with respect to ``E``.
+
+      If ``exports`` is not given, then the export environment of ``M``
+      includes a mapping from occurrence name to original name
+      for each definition within the local module. For each nested local
+      module ``N`` within ``M`` with export QEnv ``N_E``, the
+      export QEnv of ``M`` includes ``qualify(N, N_E)``.
```

Comment:
“each definition within the local module” means including definitions in nested local modules, right? E.g.
```
module Top where
  module A where
    module B where 
      foo = True

  x = (foo, B.foo) -- both valid
```
where `foo` counts as “defined withing module A”?

 I think the example in the next bullet point confirms that.

Reply:
Yes. Should this be clarified in the text?

Reply:
Not super important, but it wasn’t clear during my first reading, so maybe
```suggestion
      If ``exports`` is not given, then the export environment of ``M``
      includes a mapping from occurrence name to original name
      for each definition within the local module, directly or nested. For each nested local
      module ``N`` within ``M`` with export QEnv ``N_E``, the
      export QEnv of ``M`` includes ``qualify(N, N_E)``.
```

Actually, what about
```
module Top where
  module A where
    module qualified B where 
      foo = True

  x = (foo, B.foo)
  ```
The `foo` is a definition within `A`, so the prose says that now `foo` is in scope unqualified on the top level, but I would only expect `B.foo` to available.